README.md
# orbitalGame

Orbital Game Enging in Python

## Requirements

- pygame
- imagemagick (specifically convert

## Setting up for development

    ```bash
    pre-commit install
    pre-commit install --hook-type commit-msg
    ```

## Running

    ```bash
    python engine.py
    ```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

## [v0.3.1](https://github.com/jhugon/orbitalGame/releases/tag/v0.3.1) - 2022-08-29

- [`91583f7`](https://github.com/jhugon/orbitalGame/commit/91583f7ddd368cb9a1c93cc9c3603fddd46fb475) Merge pull request #10 from jhugon/universe_refactoring
- [`e9671e9`](https://github.com/jhugon/orbitalGame/commit/e9671e9f94584da1db0ae2f143398db8042a78d1) refactor: moved some UniverseView into MainWindow in ui.py
- [`bc5fc63`](https://github.com/jhugon/orbitalGame/commit/bc5fc63d5e1fc02ee7c50777e88d9338182b446a) refactor(universe): input event handling broken out into methods
- [`04839f6`](https://github.com/jhugon/orbitalGame/commit/04839f6ada2ecb06a1d071f6ff25e695d5baac64) refactor: broke out part of UniverseModel.getFuture into copyUniverse
- [`656a5cd`](https://github.com/jhugon/orbitalGame/commit/656a5cdcc3764fdf19e846151eefbc977ef3e0f8) refactor(universe): add docstrings to methods
- [`880e0c7`](https://github.com/jhugon/orbitalGame/commit/880e0c7f33758e8ea3d3e829d2e11c553efdfcce) refactor: add typing to UniverseCtrl
- [`1d61a57`](https://github.com/jhugon/orbitalGame/commit/1d61a576683544b8634307f1327bab7dbee13b7c) refactor: added type checking to UniverseView
- [`4908542`](https://github.com/jhugon/orbitalGame/commit/490854247c6c9ebf3e610e1d6a8d254b26ea7578) refactor: get rid of path printing in universe
- [`08aab97`](https://github.com/jhugon/orbitalGame/commit/08aab97f99f75e71ae65de9fabc0bcfb294bfca7) fix: universe with Vec2 works now; kinematics returns copies of Vec2s
- [`ef216f7`](https://github.com/jhugon/orbitalGame/commit/ef216f75f5cdec4984150bf98deb76d400524784) refactor: try to use Vec2 in getFuture, but not working
- [`5d531f6`](https://github.com/jhugon/orbitalGame/commit/5d531f628f0f9d09795ede4fc043a8ee441dac85) refactor: Merge branch main into universe_refactoring
- [`fe845f8`](https://github.com/jhugon/orbitalGame/commit/fe845f8f1dba4ed56ad0cdba29c59a48be779adc) revert(universe): reverted code to use tuples instead of Vec2
- [`b95a2f6`](https://github.com/jhugon/orbitalGame/commit/b95a2f600d706128fd0ff419bb66a9a98fc03df1) refactor: adding type annotations to UniverseModel

## [v0.3.0](https://github.com/jhugon/orbitalGame/releases/tag/v0.3.0) - 2022-08-26

- [`fe62b46`](https://github.com/jhugon/orbitalGame/commit/fe62b46bb8ebeef3e69a5f1205a72cdf7690ad3e) feat: add montage images of all of the sprites
- [`22b1c55`](https://github.com/jhugon/orbitalGame/commit/22b1c55496224c136ac900945072383263f55351) refactor: add type refs to Universe* in spaceobject.py
- [`83795ad`](https://github.com/jhugon/orbitalGame/commit/83795ada5bc6bc90572a15e4022278a153f55293) Merge pull request #12 from jhugon/futurepaths_typechecking
- [`602323a`](https://github.com/jhugon/orbitalGame/commit/602323a1c7fe55e8340d7356fda359f191c2f70a) refactor: type checking now works in futurepaths.py
- [`b36b741`](https://github.com/jhugon/orbitalGame/commit/b36b741ca365255058fe2ff1c324b980c635bb98) refactor: added broken type-checking to futurepaths

## [v0.2.2](https://github.com/jhugon/orbitalGame/releases/tag/v0.2.2) - 2022-08-24

- [`2bf59eb`](https://github.com/jhugon/orbitalGame/commit/2bf59ebb4d878538c9d594527c6c742b785a2857) Merge pull request #9 from jhugon/spaceobjectviewrefactor
- [`e5979e7`](https://github.com/jhugon/orbitalGame/commit/e5979e7c2e4b4e6eb16bfe6051a1279774b223e2) chore: update CHANGELOG with comments about #9
- [`5f5d9c9`](https://github.com/jhugon/orbitalGame/commit/5f5d9c941a35c5598b2f6284aa223e0acf4a71d7) refactor(spaceobject): finish adding annotations and comments to file
- [`60f06bf`](https://github.com/jhugon/orbitalGame/commit/60f06bf35c277ca2c567f328f33bea2ebe2ee88d) refactor: thrustDirection -> drawFlame and clearer
- [`d7303ef`](https://github.com/jhugon/orbitalGame/commit/d7303ef5a061e1e83f3a99c73a2b3c2e07f73abd) fix: errors in latest version of mypy
- [`f761e15`](https://github.com/jhugon/orbitalGame/commit/f761e157894b62865d30374b4b2256e388c4f2b0) ci: now only runs on pull_reqs and pushes to main
- [`c3e0ee0`](https://github.com/jhugon/orbitalGame/commit/c3e0ee06ef2bfc0d068aba420262aa14dd82b462) fix: SpaceObjectView.update type didnt match Sprite
- [`9466a34`](https://github.com/jhugon/orbitalGame/commit/9466a34ccc99aba1288e5b185d5b87dedefb0f09) refactor: SpaceObjectView prepare image/rect in seperate method

Finished refactoring classes in spaceobject, the SpaceObjectView constructor is
much more comprehensible, in particular.

## [v0.2.1](https://github.com/jhugon/orbitalGame/releases/tag/v0.2.1) - 2022-08-24

- [`fe31762`](https://github.com/jhugon/orbitalGame/commit/fe31762b859d85b69404ce97783d469de38d2086) Merge pull request #6 from jhugon/KinematicsModel
- [`7319782`](https://github.com/jhugon/orbitalGame/commit/7319782355913ed326a324719b961c9c97a00624) fix: now future paths work again, kinematic == fixed
- [`a457b02`](https://github.com/jhugon/orbitalGame/commit/a457b023579de31167fcc370664b5ee840a3d111) refactor: SpaceObjectModel uses kinematic class now
- [`5854e96`](https://github.com/jhugon/orbitalGame/commit/5854e969ebe72ef343c2623768ab99bdc3b99190) refactor: added tests for kinematics.py
- [`92c79b2`](https://github.com/jhugon/orbitalGame/commit/92c79b23a195b296f6e43d32260909e077d1d329) refactor: add kinematics.py

## [v0.2.0](https://github.com/jhugon/orbitalGame/releases/tag/v0.2.0) - 2022-08-23

- [`7bb7e88`](https://github.com/jhugon/orbitalGame/commit/7bb7e88f2ef4ea6d590cde6f718ebbf4daaced46) ci: added pytest to github ci action
- [`4100fa9`](https://github.com/jhugon/orbitalGame/commit/4100fa99a1a7cabe07371f0baff5f2bee16afb88) feat: Vec2 added distance isclose __eq__ and unit tests

test_utils.py
from utils import Vec2
from math import sqrt


class Test_Vec2:
    def test_add(self):
        v00 = Vec2(0.0, 0.0)
        v11 = Vec2(1.0, 1.0)
        v5m3 = Vec2(5.0, -3.0)
        assert v00 + v11 == v11
        tmp = v00 + v5m3
        assert tmp == v5m3

    def test_iadd(self):
        v00 = Vec2(0.0, 0.0)
        v5m3 = Vec2(5.0, -3.0)
        v00 += v00
        assert v00 == Vec2(0.0, 0.0)
        tmp = Vec2(-100, 54)
        tmp += v5m3
        assert tmp == Vec2(-95, 51)

    def test_mul(self):
        v5m3 = Vec2(5.0, -3.0)
        assert v5m3 * 5 == Vec2(25, -15)

    def test_rmul(self):
        v5m3 = Vec2(5.0, -3.0)
        assert 5 * v5m3 == Vec2(25, -15)

    def test_imul(self):
        tmp = Vec2(5.0, -3.0)
        tmp *= 5
        assert tmp == Vec2(25, -15)

    def test_eq(self):
        v5m3 = Vec2(5.0, -3.0)
        v11 = Vec2(1, 1)
        assert v5m3 == v5m3
        assert v11 == v11
        assert v11 != v5m3
        assert v5m3 != v11

    def test_str(self):
        v5m3 = Vec2(5.0, -3.0)
        s = str(v5m3)
        assert "5." in s
        assert "-3." in s

    def test_repr(self):
        v5m3 = Vec2(5.0, -3.0)
        s = repr(v5m3)
        assert "5." in s
        assert "-3." in s

    def test_tuple(self):
        v5m3 = Vec2(5.0, -3.0)
        assert v5m3.tuple() == (5.0, -3.0)

    def test_distance(self):
        v5m3 = Vec2(5.0, -3.0)
        assert v5m3.distance(v5m3) == 0.0
        assert v5m3.distance(Vec2(4.0, -3.0)) == 1.0
        assert v5m3.distance(Vec2(5.0, 4)) == 7.0
        assert v5m3.distance(Vec2(4.0, -2.0)) == sqrt(2)

    def test_isClose(self):
        v5m3 = Vec2(5.0, -3.0)
        assert v5m3.isClose(v5m3, 1e-9)
        assert v5m3.isClose(v5m3, 1e9)
        assert v5m3.isClose(Vec2(4.0, -3.0), 1.1)
        assert not v5m3.isClose(Vec2(4.0, -3.0), 0.9)
        assert not v5m3.isClose(Vec2(5.0, 4), 6.9)
        assert v5m3.isClose(Vec2(5.0, 4), 7.0)
        assert v5m3.isClose(Vec2(5.0, 4), 7.1)
        assert not v5m3.isClose(Vec2(4.0, -2.0), 1)
        assert v5m3.isClose(Vec2(4.0, -2.0), 2)

    def test_rotated(self):
        v5m3 = Vec2(5.0, -3.0)
        assert v5m3.rotated(180).isClose(Vec2(-5.0, 3.0), 1e-9)
        assert v5m3.rotated(90).isClose(Vec2(3.0, 5.0), 1e-9)
        assert v5m3.rotated(-90).isClose(Vec2(-3.0, -5.0), 1e-9)
        assert v5m3.rotated(0).isClose(v5m3, 1e-9)

    def test_magnitude(self):
        v5m3 = Vec2(5.0, -3.0)
        assert v5m3.magnitude() == sqrt(25 + 9)
        assert Vec2(50, 0).magnitude() == 50.0
        assert Vec2(0.0, 23).magnitude() == 23.0

    def test_normalized(self):
        v5m3 = Vec2(5.0, -3.0)
        assert v5m3.normalized() == v5m3 * (1.0 / sqrt(25 + 9))
        assert Vec2(50, 0).normalized() == Vec2(1.0, 0.0)
        assert Vec2(0.0, 23).normalized() == Vec2(0.0, 1.0)


test_kinematics.py
from kinematics import ObjectKinematics
from utils import Vec2
from math import sqrt, pi


class Test_ObjectKinematics:
    def test_getters(self):
        k = ObjectKinematics(Vec2(0.0, 0.0), Vec2(0.0, 0.0), Vec2(0.0, 0.0))
        assert k.getPosition() == Vec2(0.0, 0.0)
        assert k.getVelocity() == Vec2(0.0, 0.0)
        assert k.getAcceleration() == Vec2(0.0, 0.0)
        k = ObjectKinematics(Vec2(1.0, 2.0), Vec2(0.0, 1000.0), Vec2(5.0, 3.0))
        assert k.getPosition() == Vec2(1.0, 2.0)
        assert k.getVelocity() == Vec2(0.0, 1000.0)
        assert k.getDirection() == -pi / 2.0
        assert k.getDirectionDeg() == -90.0
        assert k.getAcceleration() == Vec2(5.0, 3.0)
        k.updateAcceleration(Vec2(3.0, 10.0))
        assert k.getAcceleration() == Vec2(3.0, 10.0)

    def test_updatePosVel(self):
        k = ObjectKinematics(Vec2(0.0, 0.0), Vec2(0.0, 0.0), Vec2(1.0, 0.0))
        assert k.getVelocity() == Vec2(0.0, 0.0)
        assert k.getPosition() == Vec2(0.0, 0.0)
        assert k.getAcceleration() == Vec2(1.0, 0.0)
        k.updatePosVel(1.0)
        assert k.getVelocity() == Vec2(1.0, 0.0)
        assert k.getPosition() == Vec2(1.0, 0.0)
        assert k.getAcceleration() == Vec2(1.0, 0.0)
        k.updatePosVel(2.0)
        assert k.getVelocity() == Vec2(3.0, 0.0)
        assert k.getPosition() == Vec2(7.0, 0.0)
        assert k.getAcceleration() == Vec2(1.0, 0.0)

        k = ObjectKinematics(Vec2(1000.0, 0.0), Vec2(0.0, 0.0), Vec2(-5.0, -10.0))
        assert k.getVelocity() == Vec2(0.0, 0.0)
        assert k.getPosition() == Vec2(1000.0, 0.0)
        assert k.getAcceleration() == Vec2(-5.0, -10.0)
        k.updatePosVel(100.0)
        assert k.getVelocity() == Vec2(-500.0, -1000.0)
        assert k.getPosition() == Vec2(1000.0 - 50000, -100000)
        assert k.getAcceleration() == Vec2(-5.0, -10.0)


ui.py
"""
Main screen/window through which the user plays the game
"""

import pygame  # type: ignore
from typing import Optional, List, Any, Tuple, TYPE_CHECKING


class MainWindow(pygame.Surface):
    """
    Main screen/window through which the user plays the game
    """

    def __init__(self, size: Tuple[int, int], backgroundImageLoc: str) -> None:
        """
        size is a tuple (x,y): the size of the layer (world) in pixels
        """
        self.screen = pygame.display.set_mode(size)
        pygame.Surface.__init__(self, size)
        self.fill((0, 0, 0))
        self.background: Optional[pygame.surface.Surface] = None
        if backgroundImageLoc is None:
            self.background = pygame.Surface(size)
            self.background.fill((0, 0, 0))
        else:
            try:
                self.background = pygame.image.load(backgroundImageLoc)
            except pygame.error as e:
                print(("Cannot load background image:", backgroundImageLoc))
                raise SystemExit(e)
            self.background = self.background.convert()
            self.background = pygame.transform.smoothscale(self.background, size)
            self.blit(self.background, (0, 0))
        self.size = size

        pygame.display.set_caption("Orbital Game")
        self.screen.blit(self, (0, 0))
        pygame.display.update()

    def update(self) -> None:
        """
        Draw Everything
        """
        if self.background is None:
            raise ValueError("background has not been set")
        self.screen.blit(self, (0, 0))


utils.py
"""
Utility and general-use functions/classes
"""

import pygame  # type: ignore
from pygame.locals import RLEACCEL  # type: ignore
import math
from math import sqrt
import os.path
from typing import Tuple, Optional

main_dir = os.path.split(os.path.abspath(__file__))[0]
sprites_dir = os.path.join(main_dir, "sprites")


def load_image(image_filename: str) -> Tuple[pygame.surface.Surface, pygame.rect.Rect]:
    """
    Load image from image_filename into pygame objects
    Converts the image to use alpha transparancy
    """
    try:
        image = pygame.image.load(image_filename)
    except pygame.error as e:
        print(("Cannot load image:", image_filename))
        raise SystemExit(e)
    image = image.convert_alpha()
    return image, image.get_rect()


class Vec2:
    """
    2D vector class useful for kinematics and geometry
    """

    ## coordinate storage
    x: float
    y: float

    def __init__(self, x: float, y: float) -> None:
        """
        x and y are the coordinates
        """
        self.x = x
        self.y = y

    def distance(self, other: "Vec2") -> float:
        """
        Distance between other and self
        """
        dx = self.x - other.x
        dy = self.y - other.y
        return sqrt(dx**2 + dy**2)

    def isClose(self, other: "Vec2", distance: float) -> bool:
        """
        Is other within distance of self?
        """
        d = self.distance(other)
        return d <= distance

    def rotate(self, angleDeg: float) -> None:
        """
        rotate this instance
        """
        angle = math.radians(angleDeg)
        x = self.x
        y = self.y
        self.x = x * math.cos(angle) - y * math.sin(angle)
        self.y = x * math.sin(angle) + y * math.cos(angle)

    def rotated(self, angleDeg: float) -> "Vec2":
        """
        return a new rotated version of this instance
        """
        result = Vec2(self.x, self.y)
        result.rotate(angleDeg)
        return result

    def tuple(self) -> Tuple[float, float]:
        """
        return vector as a tuple of floats (x,y)
        """
        return self.x, self.y

    def magnitude(self) -> float:
        """
        return sqrt(x**2+y**2)
        """
        return sqrt(self.x**2 + self.y**2)

    def normalize(self) -> None:
        """
        normalizes this vector i.e. makes it's magnitude one
        """
        mag = self.magnitude()
        self.x /= mag
        self.y /= mag

    def normalized(self) -> "Vec2":
        """
        returns a normalized version of this vector
            i.e. same direction but magnitude one
        """
        result = Vec2(self.x, self.y)
        result.normalize()
        return result

    def __add__(self, other: "Vec2") -> "Vec2":
        """
        Vector addition == translation
        """
        return Vec2(self.x + other.x, self.y + other.y)

    def __iadd__(self, other: "Vec2") -> "Vec2":
        """
        Vector addition == translation
        In-place
        """
        self.x += other.x
        self.y += other.y
        return self

    def __sub__(self, other: "Vec2") -> "Vec2":
        """
        Vector subtraction
        """
        return Vec2(self.x - other.x, self.y - other.y)

    def __isub__(self, other: "Vec2") -> "Vec2":
        """
        Vector subtraction
        In-place
        """
        self.x -= other.x
        self.y -= other.y
        return self

    def __mul__(self, sf: float) -> "Vec2":
        """
        Return this vector scaled by scale factor sf
        """
        return Vec2(self.x * sf, self.y * sf)

    def __rmul__(self, sf: float) -> "Vec2":
        """
        Return this vector scaled by scale factor sf
        """
        return self * sf

    def __imul__(self, sf: float) -> "Vec2":
        """
        Scale this vector by scale factor sf
        """
        self.x *= sf
        self.y *= sf
        return self

    def __eq__(self, other) -> bool:
        """
        Equality of underlying coordinates
        """
        return (self.x == other.x) and (self.y == other.y)

    def __str__(self) -> str:
        return f"Vec2({self.x},{self.y})"

    def __repr__(self) -> str:
        return str(self)

    def copy(self) -> "Vec2":
        return Vec2(self.x, self.y)


engine.py
#!/usr/bin/python

from math import sqrt
import pygame  # type: ignore
from spaceobject import SpaceObjectCtrl
from universe import UniverseCtrl


class SpaceApplication:
    def __init__(self):
        pygame.init()
        windowsize = (800, 600)
        backgroundImageLoc = (
            "backgroundExt/night-sky-milky-way-galaxy-astrophotography_0p25.jpg"
        )
        universe = UniverseCtrl(windowsize, backgroundImageLoc)

        mEarth = 6.0e24  # kg
        rVehicle = 3.5e7  # meters
        G = 6.67e-11  # in mks
        vVehicle = sqrt(G * mEarth / rVehicle)  # For circular orbit
        earth = SpaceObjectCtrl(universe, "sprites/planet3.png", 0.5, 0.0, 0.0, mEarth)
        vehicle = SpaceObjectCtrl(
            universe, "sprites/FighterLaser_springgreen.png", 1.0, rVehicle, 0.0
        )
        vehicle2 = SpaceObjectCtrl(
            universe, "sprites/SatelliteBase16_red.png", 1.0, 0.0, rVehicle
        )
        vehicle3 = SpaceObjectCtrl(
            universe, "sprites/FrigateMissile_cyan.png", 1.0, 0.0, -rVehicle
        )
        vehicle.model.kinematics.velocity.y = vVehicle
        vehicle2.model.kinematics.velocity.x = vVehicle
        vehicle3.model.kinematics.velocity.x = -vVehicle

        universe.run()


if __name__ == "__main__":
    sa = SpaceApplication()


universe.py
"""
Universe controls all of the SpaceObjects
"""

import pygame  # type: ignore
from pygame.locals import QUIT, KEYUP, KEYDOWN, K_ESCAPE, K_UP, K_DOWN, MOUSEBUTTONUP, MOUSEBUTTONDOWN  # type: ignore
from math import sqrt
from copy import deepcopy
from typing import Optional, List, Any, Tuple, TYPE_CHECKING

from utils import Vec2
from futurepaths import FuturePathsView
from spaceobject import SpaceObjectModel, SpaceObjectCtrl, SpaceObjectView
from ui import MainWindow

if TYPE_CHECKING:
    from spaceobject import SpaceObjectModel, SpaceObjectView, SpaceObjectCtrl


class UniverseModel:
    """
    Models the dynamics of the universe of SpaceObjectModels
    """

    def __init__(self, G: float = 6.67e-11, rPower: float = -2.0) -> None:
        """
        G is the gravitational constant
        rPower is the power of gravity, e.g. a = G*M*r^(rPower)
        """
        self.massiveObjects: List[SpaceObjectModel] = []
        self.masslessObjects: List[SpaceObjectModel] = []
        self.G: float = G
        self.rPower: float = rPower

    def addObject(self, obj: SpaceObjectModel) -> None:
        obj.universe = self
        if obj.mass > 0.0:
            self.massiveObjects += [obj]
        else:
            self.masslessObjects += [obj]

    def getA(self, position: Vec2) -> Vec2:
        """
        Get the gravitational acceleration at a point in space
        """
        acceleration = Vec2(0.0, 0.0)
        for mo in self.massiveObjects:
            rVec = mo.kinematics.getPosition() - position
            r = rVec.magnitude()
            if r < 0.001:
                continue
            accmagfrommo = self.G * mo.mass * r ** (self.rPower)
            accfrommo = accmagfrommo * rVec.normalized()
            acceleration += accfrommo
        return acceleration

    def update(self, dt: float) -> None:
        """
        Update all of the objects' acceleration, velocity, position, and thrusts
        """
        for obj in self.massiveObjects + self.masslessObjects:
            obj.update1(dt)
        for obj in self.massiveObjects + self.masslessObjects:
            obj.update2(dt)

    def __str__(self) -> str:
        result = ""
        for obj in self.massiveObjects + self.masslessObjects:
            result += str(obj)
        return result

    def getFuture(
        self,
        dtList: List[float],
        selectedObj: Optional[SpaceObjectModel] = None,
        dtStepSize: float = 1e2,
    ) -> Tuple[List[List[Vec2]], List[List[float]]]:
        """
        Get the future positions and thrusts of all massless objects in the universe

        dtStepSize is in model seconds, just like dtList
        """
        futureUniverse, mlos = self.copyUniverse(selectedObj)

        futurePositionList: List[List[Vec2]] = [[] for i in mlos]
        futureBurnList: List[List[float]] = [[] for i in mlos]
        iDt = 0
        dtTotal = 0.0
        while True:
            dt = dtList[iDt]
            dtStep = dtStepSize
            recordThisStep = False
            if dt - dtTotal < dtStepSize:
                dtStep = dt - dtTotal
                recordThisStep = True
            futureUniverse.update(dtStep)
            if recordThisStep:
                for i in range(len(mlos)):
                    pos = mlos[i].kinematics.getPosition()
                    futurePositionList[i] += [pos]
                    burn = 0.0
                    for burnTime in mlos[i].burnSchedule:
                        if burnTime[0] <= 0.0:
                            burn += burnTime[2]
                    futureBurnList[i] += [burn]
                iDt += 1
            dtTotal += dtStep
            if iDt >= len(dtList):
                break
        return futurePositionList, futureBurnList

    def copyUniverse(
        self, selectedObj: Optional["SpaceObjectModel"] = None
    ) -> Tuple["UniverseModel", List[SpaceObjectModel]]:
        """
        Make a copy of this universe and also return a list of its massless objects.
        If selectedObj is not None, then make sure it is first in the list
        """
        futureUniverse = deepcopy(self)
        mlos = futureUniverse.masslessObjects
        if selectedObj is not None:
            foundSelected = False
            mlosNew: List[SpaceObjectModel] = []
            for obj in mlos:
                if obj.kinematics == selectedObj.kinematics:
                    selectedObj = obj
                    foundSelected = True
                    continue
                mlosNew += [obj]
            mlos = mlosNew
            if foundSelected:
                mlos.append(selectedObj)
                mlos.reverse()
            assert foundSelected
        return futureUniverse, mlos


######################################################3


class UniverseView:
    def __init__(self, window: "MainWindow") -> None:
        self.window = window
        self.objects: pygame.sprite.RenderUpdates = pygame.sprite.RenderUpdates()
        self.selected = pygame.sprite.Group()
        self.hudGroup = pygame.sprite.RenderUpdates()
        self.toUpdateRectsList: List[pygame.rect.Rect] = []

    def addObject(self, obj: "SpaceObjectView") -> None:
        obj.setUniverse(self)
        self.objects.add(obj)

    def update(self) -> None:
        """
        Update everything
        """
        self.objects.update()
        self.toUpdateRectsList += self.objects.draw(self.window.screen)
        self.toUpdateRectsList += self.hudGroup.draw(self.window.screen)

        pygame.display.update(self.toUpdateRectsList)  # type: ignore
        self.toUpdateRectsList = []

    def deselectAll(self) -> None:
        """
        Make sure no objects are selected
        """
        for obj in self.objects:
            if not hasattr(obj, "deSelect"):
                raise TypeError(
                    "object isn't an instance of SpaceObjectView or otherwise doesn't have deSelect method"
                )
            obj.deSelect()  # type: ignore

    def showPaths(
        self,
        futurePaths: List[List[Tuple[int, int]]],
        futureBurns: List[List[float]],
        timePoints: Optional[List[float]] = None,
        selected: bool = False,
    ) -> None:
        """
        Draw the future paths, highlighting the first entry in the list, if selected is True
        """
        pathsView = FuturePathsView(self)
        selectedBools = [False for i in range(len(futurePaths))]
        selectedBools[0] = selected
        for objPath, objBurns, selectedPathBool in reversed(
            list(zip(futurePaths, futureBurns, selectedBools))
        ):
            pathsView.addPath(selectedPathBool, objPath, objBurns, timePoints)


######################################################3


class UniverseCtrl:
    def __init__(
        self, size: Tuple[int, int], backgroundImageLoc: str, debug: bool = False
    ) -> None:
        """
        Controls the program

        size is a tuple (x,y): the size of the layer (world) in pixels
        """
        self.viewSize = size
        self.debug = debug
        modelSize = 3.5e7 * 3.0
        # self.meterPerPixel = 1e10
        self.meterPerPixel = modelSize / self.viewSize[1]
        self.speedUpFactor = 5e3
        self.updateModelEvery = 1e2  # seconds of model time
        self.dRClickPath = 25.0

        self.mainwindow = MainWindow(size, backgroundImageLoc)
        self.model = UniverseModel()
        self.view = UniverseView(self.mainwindow)
        self.objects: List[SpaceObjectCtrl] = []

        self.selected: List[SpaceObjectCtrl] = []

        self.pauseModel = False
        self.selectedModeFlag = False

        self.selectedPathPointsView: Optional[List[Tuple[int, int]]] = None
        self.selectedPathTimes: Optional[List[float]] = None
        self.selectedBurnStartIndex: Optional[int] = None

    def addObject(self, obj: "SpaceObjectCtrl") -> None:
        self.objects += [obj]
        self.model.addObject(obj.model)
        self.view.addObject(obj.view)

    def convertCoordsModel2View(self, x: float, y: float) -> Tuple[int, int]:
        """
        Convert from model/dynamics coordinates to view/screen/window coordinates
        """
        return (
            int(x // self.meterPerPixel + self.viewSize[0] // 2),
            int(-y // self.meterPerPixel + self.viewSize[1] // 2),
        )

    def convertCoordsView2Model(self, x: int, y: int) -> Tuple[float, float]:
        """
        Convert from view/screen/window coordinates to model/dynamics coordinates
        """
        return (x - self.viewSize[0] / 2) * self.meterPerPixel, (
            y - self.viewSize[1] / 2
        ) * self.meterPerPixel

    def updateViewToModel(self) -> None:
        """
        Make sure the view/screen/window matches the model/dynamics
        """
        for obj in self.objects:
            obj.updateViewToModel()

    def run(self) -> None:
        """
        Start the game
        """
        clock = pygame.time.Clock()
        running = True
        counter = 0.0
        while running:
            clock.tick(60)
            dt = clock.get_time() / 1000.0  # Convert from ms to s
            counter += dt
            if counter > 2.0:
                if self.debug:
                    print(("fps: {0}".format(clock.get_fps())))
                    print((self.model))
                counter = 0.0

            # Handle Input Events
            for event in pygame.event.get():
                running = running and self.handleUIEvents(event)

            # Update Model
            if not self.pauseModel:
                dtModel = dt * self.speedUpFactor
                nModelUpdates = int(dtModel / self.updateModelEvery)
                dtRemainder = dtModel % self.updateModelEvery
                for i in range(nModelUpdates):
                    self.model.update(self.updateModelEvery)
                self.model.update(dtRemainder)
            # Update View to model
            self.updateViewToModel()

            # Update View
            self.mainwindow.update()
            self.view.update()

        ## End of event loop
        pygame.quit()

    def handleUIEvents(self, event: pygame.event.Event) -> bool:
        running = True
        if event.type == QUIT:
            running = False
        elif event.type == KEYDOWN and event.key == K_ESCAPE:
            running = False

        elif event.type == KEYDOWN and event.key == K_UP:
            for obj in self.selected:
                obj.model.thrust = 1.0
        elif event.type == KEYDOWN and event.key == K_DOWN:
            for obj in self.selected:
                obj.model.thrust = -1.0
        elif event.type == KEYUP and (event.key == K_UP or event.key == K_DOWN):
            for obj in self.objects:
                obj.model.thrust = 0.0
        elif event.type == MOUSEBUTTONDOWN and event.dict["button"] == 1:
            self.handleMouseButtonDownEvent(event)
        elif event.type == MOUSEBUTTONUP:
            self.handleMouseButtonUpEvent(event)
        return running

    def handleMouseButtonDownEvent(self, event: pygame.event.Event) -> None:
        assert event.type == MOUSEBUTTONDOWN and event.dict["button"] == 1
        mousePosition = event.dict["pos"]
        clickedObjects = self.findObjectsAtPoint(mousePosition)
        nClickedObjects = len(clickedObjects)
        if nClickedObjects == 1:
            self.selectObject(clickedObjects[0])
        elif self.selectedModeFlag:
            startIndex = self.isCloseToFuturePath(mousePosition)
            if startIndex is not None:
                self.selectedBurnStartIndex = startIndex
            else:
                self.deselectAll()
        elif nClickedObjects == 0:
            self.deselectAll()

    def handleMouseButtonUpEvent(self, event: pygame.event.Event) -> None:
        assert event.type == MOUSEBUTTONUP
        mousePosition = event.dict["pos"]
        if (
            self.selectedModeFlag
            and self.selectedBurnStartIndex is not None
            and self.selectedPathTimes is not None
        ):
            endIndex = self.isCloseToFuturePath(mousePosition)
            if endIndex is not None and endIndex != self.selectedBurnStartIndex:
                burnStartT = self.selectedPathTimes[self.selectedBurnStartIndex]
                burnEndT = self.selectedPathTimes[endIndex]
                if burnEndT is not None:
                    if burnEndT > burnStartT:
                        self.selected[0].scheduleBurn(burnStartT, burnEndT, 1.0)
                    if burnEndT < burnStartT:
                        self.selected[0].scheduleBurn(burnEndT, burnStartT, -1.0)
            else:
                self.selectedBurnStartIndex = None

    def findObjectsAtPoint(self, point: Tuple[int, int]) -> List["SpaceObjectCtrl"]:
        """
        Find an object (if any) at the given (view/screen) point
        """
        result: List["SpaceObjectCtrl"] = []
        for obj in self.objects:
            if obj.view.rect.collidepoint(*point):
                result += [obj]
        return result

    def selectObject(self, selectedObject: "SpaceObjectCtrl") -> None:
        """
        Select a space object, and bring the UI into selected mode
        """
        self.deselectAll()
        selectedObject.select()
        self.pauseModel = True
        self.selectedModeFlag = True
        self.showPaths()

    def deselectAll(self) -> None:
        """
        Make sure no space object is selected and the UI isn't in selected mode
        """
        self.pauseModel = False
        self.selectedModeFlag = False
        for obj in self.objects:
            obj.selected = False
        self.view.deselectAll()
        self.view.hudGroup.empty()
        self.selected = []
        self.selectedPathPointsView = None
        self.selectedPathTimes = None
        self.selectedBurnStartIndex = None

    def showPaths(self) -> None:
        """
        Show space object paths in view
        """
        self.view.hudGroup.empty()
        timePoints = [i * 1e3 for i in range(30)]
        selectedModel: Optional[SpaceObjectModel] = self.selected[0].model
        if selectedModel is None or selectedModel.mass > 0.0:
            selectedModel = None
        futurePaths, futureBurns = self.model.getFuture(
            timePoints, selectedObj=selectedModel
        )
        futurePathsView: List[List[Tuple[int, int]]] = []
        for path in futurePaths:
            pathView: List[Tuple[int, int]] = []
            for p in path:
                pView = self.convertCoordsModel2View(*(p.tuple()))
                pathView += [pView]
            futurePathsView += [pathView]
        selected = True
        if selectedModel is None or selectedModel.mass > 0.0:
            selected = False
        self.view.showPaths(futurePathsView, futureBurns, timePoints, selected=selected)

        if selectedModel is not None and selectedModel.mass == 0.0:
            self.selectedPathPointsView = futurePathsView[0]
            self.selectedPathTimes = timePoints

    def isCloseToFuturePath(self, pos: Tuple[int, int]) -> Optional[int]:
        """
        Check if position is close to a path
        """
        if self.selectedPathPointsView is None:
            return None
        drMax2 = self.dRClickPath**2
        x, y = pos
        dR2List = [
            (xS - x) ** 2 + (yS - y) ** 2 for xS, yS in self.selectedPathPointsView
        ]
        minR2 = min(dR2List)
        result = minR2 < drMax2
        if not result:
            return None
        # print("minR: {0:.1f}, result: {1}".format(sqrt(minR2),result))
        iMinR2 = dR2List.index(minR2)
        return iMinR2
